management/WFCORE-1110_Custom_HTTP_Management_Headers.adoc

```diff
@@ -0,0 +1,84 @@
+= <INSERT TITLE HERE>
+:author:            Darran Lofthouse
+:email:             darran.lofthouse@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFCORE-1110[WFCORE-1110]
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/WFLY-12787[WFLY-12787]
+* https://issues.jboss.org/browse/EAP7-1335[EAP7-1335]
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+=== Testing By
+// Put an x in the relevant field to indicate if testing will be done by Engineering or QE. 
+// Discuss with QE during the Kickoff state to decide this
+[ ] Engineering
+
+[X] QE
+
+=== Affected Projects or Components
+
+=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+
+The HTTP management endpoint of the application server already returns a pre-defined set of HTTP headers in all of the responses that are sent to the clients, however it is sometimes desirable that additional headers can be returned to satisfy the requirements of an environment the application server is installed within.  This enhancement is to add support for an end user to specify a custom set of HTTP headers that should be returned within each response.
+
+Predominantly we have two endpoints made available from the HTTP management endpoint by default.
+
+ * `/management` - The endpoint for servicing management requests over HTTP.
+ * `/console` - The endpoint to provide the admin console initial page and complete JavaScript.
+
+As we have different endpoints it may be desirable for them to return different headers, as an example some headers are only applicable to the handling of pages or interpretation of the JavaScript so should only be applied to the `/console` endpoint whilst other headers may be applicable to both.
+
+Configuration will be added to the HTTP management endpoint to allow headers to be specified, an example configuration will be added later but in general the following will be apply: -
+
+ * The configuration will allow the end user to specify which context the headers apply to or if they apply to both contexts.
+ * This will be a new `custom-headers` attribute on the `/core-service=management/management-interface=http-interface` resource.
+ * Within the `custom-headers` attribute it will be possible to specify a set of headers represented as an ordered list of key / value pairs against a specific patch.
```

Comment:
s/patch/path

Reply:
So each element in the list would consist of a node encapsulating

1) the relevant path
2) header name
3) header value

?

The path being any possible path, starting with "/" to cover all contexts, and then going beyond, e.g.  a specific context or even /management/subsystem/logging/console-handler.  But nothing fancy like regex matching.

(I'm not advocating for /management/subsystem/logging/console-handler, although doing it is fine if it's easier. I'm just clarifying.)

Reply:
I will add a clarification for this I think we will restrict it to: -

- /
- /console
- /management

Keeping the options simple, at a later point deeper paths / redgex etc could be considered but TBH I think the root contexts will be sufficient - at this level we pick up that one handles JavaScript and once handles API calls with an option to apply headers to both equally.

Reply:
Sounds good.

Reply:
Should the `/health` and `/metrics` also be covered?

Looking at the implementation, that's actually the case

Reply:
Yes all paths below `/` are covered.